### PR TITLE
HW_03 by Ivan_Papushoi

### PR DESCRIPTION
**Pull request by 18bitmood**
Created new PR instead previous

### CODE REVIEW

File: HW03/Ivan_Papushoi/app.rb [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    puts
+  end
+
+  def print_students_submitted(student)
+    create_str_students_submitted(student) unless student.submitted_hws.empty?
+  end
+
+  def create_str_students_submitted(student)
+    print "\n#{student.name} - ##{student.submitted_hws.join ', '} "
+  end
```
Comment (B0RlS) [inline HW03/Ivan_Papushoi/app.rb:68]: Think about the name of the methods and their tasks, I don’t understand their meaning

File: HW03/Ivan_Papushoi/app.rb [hunk 2]
```diff
@@ -0,0 +1,156 @@
+# The core idea of the app:
+# there are objects "tasks" that assign to students and they should solve it
+# mentor see results, check it, and set statuses 'success' and 'denied'
+# if it was denied, students try to resolve it and sends results again.
+# Mentors and students can requests tasks selections
+# which were created, passed, solved and denied
+
+class Student
+  attr_reader :name, :submitted_hws, :denied_hws, :passed_hws
+  attr_writer :submitted_hws, :denied_hws, :passed_hws
```
Comment (B0RlS) [inline HW03/Ivan_Papushoi/app.rb:10]: Think about variable names, don't use abbreviations

File: HW03/Ivan_Papushoi/app.rb [hunk 3]
```diff
@@ -0,0 +1,156 @@
+# The core idea of the app:
+# there are objects "tasks" that assign to students and they should solve it
+# mentor see results, check it, and set statuses 'success' and 'denied'
+# if it was denied, students try to resolve it and sends results again.
+# Mentors and students can requests tasks selections
+# which were created, passed, solved and denied
+
+class Student
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Restructured app in few files

File: HW03/Ivan_Papushoi/app.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Use attr_accessor now

File: HW03/Ivan_Papushoi/app.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Thinked and removed, renamed, rewrited methods - now it notifies mentors, not "space"

File: HW03/Ivan_Papushoi/app.rb [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+  def initialize(name)
+    @name = name
+    @submitted_hws = []
+    @denied_hws = []
+    @passed_hws = []
+  end
+
+  def submit_homework(task_id)
+    puts "#{name} submit homework ##{task_id}"
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Removed commands into another method

File: HW03/Ivan_Papushoi/app.rb [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    @submitted_hws << task_id
+  end
+
+  def resolve_homework(task_id)
+    puts "#{name} resolve homework ##{task_id}"
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Same solution

File: HW03/Ivan_Papushoi/app.rb [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  end
+end
+
+class Task
+  attr_reader :task_id, :name
+  attr_writer :name
+
+  def initialize(task_id, students_array)
+    @task_id = task_id
+    puts "Task #{@task_id} was created"
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Fixed

File: HW03/Ivan_Papushoi/app.rb [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    notify_students(@task_id, students_array)
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Fixed

File: HW03/Ivan_Papushoi/app.rb [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  end
+
+  def notify_students(task_id, students_array)
+    notify_string = " was notified about task #{task_id}"
+    students_array.each { |student| puts student.name + notify_string }
+  end
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Now task doesn't notify anyone

File: HW03/Ivan_Papushoi/app.rb (hunk 1 above)
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Fixed this

File: HW03/Ivan_Papushoi/app.rb (hunk 2 above)
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: I see, it has low readability. Fixed.

File: HW03/Ivan_Papushoi/app.rb [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    puts "#{name} resolve homework ##{task_id}"
+    @denied_hws.delete(task_id)
+    @submitted_hws << task_id
+  end
+
+  def check_submitted
+    print "#{name} check. Next homeworks submitted: "
+    @submitted_hws.each { |x| print "##{x}  " }
+    puts
+  end
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Deleted that strange method

File: HW03/Ivan_Papushoi/app.rb [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+  def set_status_fail(student, task_id)
+    student.submitted_hws.delete(task_id)
+    student.denied_hws.push(task_id)
+    puts "Mentor set status -denied- for #{student.name} task ##{task_id}"
+  end
+end
+
+class Task
+  attr_reader :task_id, :name
+  attr_writer :name
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Fixed

File: HW03/Ivan_Papushoi/mentor.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
+class Mentor
+  attr_accessor :name, :students_array
+
+  def initialize(name)
+    @name = name
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb:5]: Fixed, thank you for advice

File: HW03/Ivan_Papushoi/mentor.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
+class Mentor
+  attr_accessor :name, :students_array
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb]: Now it is 'students'

File: HW03/Ivan_Papushoi/mentor.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+  def initialize(name)
+    @name = name
+  end
+
+  def create_task(task_id)
+    Task.new(task_id)
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb:10]: I don't have a class called 'homework' in my code, just tasks, so I can't pass it

File: HW03/Ivan_Papushoi/mentor.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    notify_students(task_id)
+  end
+
+  def notify_students(task_id)
+    message = " was notified about task #{task_id}"
+    @students_array.each { |student| puts 'Student ' + student.name + message }
+  end
+
+  def check_students_submitted
+    print 'Mentor checks. Next students submitted their homeworks:  '
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb]: Fixed

File: HW03/Ivan_Papushoi/mentor.rb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    @students_array.each(&:print_submitted_homework)
+    puts
+  end
+
+  def set_status_success(student, task_id)
+    student.submitted_homeworks.delete(task_id)
+    student.passed_homeworks.push(task_id)
+    puts "Mentor set status -success- for #{student.name} task ##{task_id}"
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb]: Fixed

File: HW03/Ivan_Papushoi/mentor.rb [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  def set_status_fail(student, task_id)
+    student.submitted_homeworks.delete(task_id)
+    student.denied_homeworks.push(task_id)
+    puts "Mentor set status -denied- for #{student.name} task ##{task_id}"
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb]: Fixed

File: HW03/Ivan_Papushoi/student.rb [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    @submitted_homeworks << task_id
+    notify_mentor(task_id)
+  end
+
+  def notify_mentor(task_id)
+    puts "#{name} submit homework ##{task_id}"
+  end
+
+  def check_denied
+    print "#{name} check. Next homeworks denied: "
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/student.rb]: Fixed

File: HW03/Ivan_Papushoi/student.rb [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    @denied_homeworks.each { |x| print "##{x}  " }
+    puts
+  end
+
+  def check_passed
+    print "#{name} check. Next homeworks passed: "
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/student.rb]: Fixed

File: HW03/Ivan_Papushoi/student.rb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    @passed_homeworks.each { |x| print "##{x}  " }
+    puts
+  end
+
+  def print_submitted_homework
+    submitted = "\n#{name} - ##{submitted_homeworks.join ', '} "
+    print submitted if submitted_homeworks.any?
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/student.rb]: Fixed

**Comment by 18bitmood** [comment]
Fixed issues

File: HW03/Ivan_Papushoi/mentor.rb [hunk 20]
```diff
@@ -0,0 +1,36 @@
+class Mentor
+  attr_accessor :name, :students_array
+
+  def initialize(name)
+    @name = name
+  end
+
+  def create_task(task_id)
+    Task.new(task_id)
+    puts "Mentor #{name} create task ##{task_id}"
```
[the remark this answers is not in the document]
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb]: Fixed

File: HW03/Ivan_Papushoi/app.rb [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+# Mentors and students can requests tasks selections
+# which were created, passed, solved and denied
+
+require_relative 'student'
+require_relative 'mentor'
+require_relative 'task'
+
+students = []
+students << student1 = Student.new('Igor')
+students << student2 = Student.new('Sergey')
```
Comment (vladhilko) [inline HW03/Ivan_Papushoi/app.rb:14]: Would be better to go separately 
```suggestion
students = []
student1 = Student.new('Igor')
student2 = Student.new('Sergey')
students << student1
students << student2
```

File: HW03/Ivan_Papushoi/student.rb [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+  def check_denied
+    denied = ''
+    @denied_homeworks.each { |x| denied += "##{x} " }
+    "#{name} check. Next homeworks denied: " + denied
+  end
+
+  def check_passed
+    passed = ''
+    @passed_homeworks.each { |x| passed += "##{x}  " }
+    "#{name} check. Next homeworks passed: " + passed
```
Comment (vladhilko) [inline HW03/Ivan_Papushoi/student.rb:36]:
```suggestion
    "#{name} check. Next homeworks passed: " + passed_homeworks.join('# ')
```

File: HW03/Ivan_Papushoi/student.rb [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def notify_mentor(task_id)
+    "#{name} submit homework ##{task_id}"
+  end
+
+  def check_denied
+    denied = ''
+    @denied_homeworks.each { |x| denied += "##{x} " }
+    "#{name} check. Next homeworks denied: " + denied
```
Comment (vladhilko) [inline HW03/Ivan_Papushoi/student.rb:30]: the same

File: HW03/Ivan_Papushoi/mentor.rb [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+  def create_task(task_id)
+    Task.new(task_id)
+    "Mentor #{name} create task ##{task_id} \n" + notify_students(task_id)
+  end
+
+  def notify_students(task_id)
+    message = " was notified about task #{task_id} \n"
+    notify = ''
+    @students.each { |student| notify += 'Student ' + student.name + message }
+    notify
```
Comment (vladhilko) [inline HW03/Ivan_Papushoi/mentor.rb:18]: Could you implement this logic on the `Student` class level?
```suggestion
    students.each { |student| student.notify(task_id) }
```

File: HW03/Ivan_Papushoi/mentor.rb [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def check_students_submitted
+    submitted = ''
+    @students.each do |student|
+      submitted += student.submitted unless student.submitted.nil?
+    end
+    'Mentor checks. Next students submitted their homeworks:' + submitted
```
Comment (vladhilko) [inline HW03/Ivan_Papushoi/mentor.rb:26]: the same

File: HW03/Ivan_Papushoi/app.rb (hunk 21 above)
Comment (18bitmood) [reply HW03/Ivan_Papushoi/app.rb]: Done

File: HW03/Ivan_Papushoi/student.rb (hunk 22 above)
Comment (18bitmood) [reply HW03/Ivan_Papushoi/student.rb]: Thank you for advice. Changed

File: HW03/Ivan_Papushoi/student.rb (hunk 23 above)
Comment (18bitmood) [reply HW03/Ivan_Papushoi/student.rb]: Fixed

File: HW03/Ivan_Papushoi/mentor.rb (hunk 24 above)
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb]: I implemented, looks better

File: HW03/Ivan_Papushoi/mentor.rb (hunk 25 above)
Comment (18bitmood) [reply HW03/Ivan_Papushoi/mentor.rb]: Done

**Comment by 18bitmood** [comment]
Fixed issues from review

**Pull request merged**